src/cli/rustup_mode.rs

```diff
@@ -136,536 +129,577 @@ pub fn main() -> Result<utils::ExitCode> {
+            Some(("set", m)) => override_add(cfg, m)?,
+            Some(("unset", m)) => override_remove(cfg, m)?,
+            _ => unreachable!(),
         },
-        ("run", Some(m)) => run(cfg, m)?,
-        ("which", Some(m)) => which(cfg, m)?,
-        ("doc", Some(m)) => doc(cfg, m)?,
-        ("man", Some(m)) => man(cfg, m)?,
-        ("self", Some(c)) => match c.subcommand() {
-            ("update", Some(_)) => self_update::update(cfg)?,
-            ("uninstall", Some(m)) => self_uninstall(m)?,
-            (_, _) => unreachable!(),
+        Some(("run", m)) => run(cfg, m)?,
+        Some(("which", m)) => which(cfg, m)?,
+        Some(("doc", m)) => doc(cfg, m)?,
+        Some(("man", m)) => man(cfg, m)?,
+        Some(("self", c)) => match c.subcommand() {
+            Some(("update", _)) => self_update::update(cfg)?,
+            Some(("uninstall", m)) => self_uninstall(m)?,
+            _ => unreachable!(),
         },
-        ("set", Some(c)) => match c.subcommand() {
-            ("default-host", Some(m)) => set_default_host_triple(cfg, m)?,
-            ("profile", Some(m)) => set_profile(cfg, m)?,
-            ("auto-self-update", Some(m)) => set_auto_self_update(cfg, m)?,
-            (_, _) => unreachable!(),
+        Some(("set", c)) => match c.subcommand() {
+            Some(("default-host", m)) => set_default_host_triple(cfg, m)?,
+            Some(("profile", m)) => set_profile(cfg, m)?,
+            Some(("auto-self-update", m)) => set_auto_self_update(cfg, m)?,
+            _ => unreachable!(),
         },
-        ("completions", Some(c)) => {
-            if let Some(shell) = c.value_of("shell") {
+        Some(("completions", c)) => {
+            if let Some(&shell) = c.get_one::<Shell>("shell") {
                 (output_completion_script(
-                    shell.parse::<Shell>().unwrap(),
-                    c.value_of("command")
-                        .and_then(|cmd| cmd.parse::<CompletionCommand>().ok())
+                    shell,
+                    c.get_one::<CompletionCommand>("command")
+                        .copied()
                         .unwrap_or(CompletionCommand::Rustup),
                 ))?
             } else {
                 unreachable!()
             }
         }
-        (_, _) => unreachable!(),
+        _ => unreachable!(),
     })
 }
 
-pub(crate) fn cli() -> App<'static, 'static> {
-    let mut app = App::new("rustup")
+pub(crate) fn cli() -> Command<'static> {
+    let mut app = Command::new("rustup")
         .version(common::version())
         .about("The Rust toolchain installer")
         .after_help(RUSTUP_HELP)
-        .setting(AppSettings::VersionlessSubcommands)
-        .setting(AppSettings::DeriveDisplayOrder)
-        .setting(AppSettings::SubcommandRequiredElseHelp)
+        .global_setting(AppSettings::DeriveDisplayOrder)
+        .arg_required_else_help(true)
+        .next_help_heading("FLAGS")
```

Comment:
Unsure about this: in `clap v2` there was [`FLAGS`, `OPTIONS`, `ARGS` sections, controlled by `clap`](https://github.com/clap-rs/clap/blob/master/CHANGELOG.md#highlights-1:~:text=You%20can%20now%20override%20the%20default%20group%20with%20Arg%3A%3Ahelp_heading%20and%20App%3A%3Asubcommand_help_heading.%20To%20apply%20a%20heading%20to%20a%20series%20of%20arguments%2C%20you%20can%20set%20App%3A%3Ahelp_heading.). Now, automatically added `-h, --help` and `-V, --version` flags are always displayed under `OPTIONS` section and there is no way to control that.

Reply:
Could you please test it and show the difference in the comment? Sorry for the delayed reply.

Reply:
```diff 
@@ -1,16 +1,18 @@
-$ rustup --help
+$ ./target/debug/rustup --help
-rustup 1.25.1 (bb60b1e89 2022-07-12)
+rustup unknown (66def540e 2022-10-15) dirty 3 modifications
 The Rust toolchain installer
 
 USAGE:
-    rustup [FLAGS] [+toolchain] <SUBCOMMAND>
+    rustup [OPTIONS] [+toolchain] [SUBCOMMAND]
+
+OPTIONS:
+    -h, --help       Print help information
+    -V, --version    Print version information
 
 FLAGS:
     -v, --verbose    Enable verbose output
     -q, --quiet      Disable progress output
-    -h, --help       Prints help information
-    -V, --version    Prints version information
 
 ARGS:
     <+toolchain>    release channel (e.g. +stable) or custom toolchain to set override
@@ -31,7 +33,7 @@ SUBCOMMANDS:
     self           Modify the rustup installation
     set            Alter rustup settings
     completions    Generate tab-completion scripts for your shell
-    help           Prints this message or the help of the given subcommand(s)
+    help           Print this message or the help of the given subcommand(s)
 
 DISCUSSION:
     Rustup installs The Rust Programming Language from the official
@@ -43,33 +45,35 @@ DISCUSSION:
     If you are new to Rust consider running `rustup doc --book` to
     learn Rust.
 
-- snip --
```

<details>
<summary>Old vs New</summary>

<table>
<tr>
	<td>Old (`rustup 1.25 stable`)</td>
	<td>New (this `PR` 66def540)</td>
</tr>
<tr>
	<td> 

```
rustup 1.25.1 (bb60b1e89 2022-07-12)
The Rust toolchain installer

USAGE:
    rustup [FLAGS] [+toolchain] <SUBCOMMAND>

FLAGS:
    -v, --verbose    Enable verbose output
    -q, --quiet      Disable progress output
    -h, --help       Prints help information
    -V, --version    Prints version information

ARGS:
    <+toolchain>    release channel (e.g. +stable) or custom toolchain to set override

SUBCOMMANDS:
    show           Show the active and installed toolchains or profiles
    update         Update Rust toolchains and rustup
    check          Check for updates to Rust toolchains and rustup
    default        Set the default toolchain
    toolchain      Modify or query the installed toolchains
    target         Modify a toolchain's supported targets
    component      Modify a toolchain's installed components
    override       Modify directory toolchain overrides
    run            Run a command with an environment configured for a given toolchain
    which          Display which binary will be run for a given command
    doc            Open the documentation for the current toolchain
    man            View the man page for a given command
    self           Modify the rustup installation
    set            Alter rustup settings
    completions    Generate tab-completion scripts for your shell
    help           Prints this message or the help of the given subcommand(s)

DISCUSSION:
    Rustup installs The Rust Programming Language from the official
    release channels, enabling you to easily switch between stable,
    beta, and nightly compilers and keep them updated. It makes
    cross-compiling simpler with binary builds of the standard library
    for common platforms.
```

</td>
<td>


```
rustup unknown (66def540e 2022-10-15) dirty 3 modifications
The Rust toolchain installer

USAGE:
    rustup [OPTIONS] [+toolchain] [SUBCOMMAND]

OPTIONS:
    -h, --help       Print help information
    -V, --version    Print version information

FLAGS:
    -v, --verbose    Enable verbose output
    -q, --quiet      Disable progress output

ARGS:
    <+toolchain>    release channel (e.g. +stable) or custom toolchain to set override

SUBCOMMANDS:
    show           Show the active and installed toolchains or profiles
    update         Update Rust toolchains and rustup
    check          Check for updates to Rust toolchains and rustup
    default        Set the default toolchain
    toolchain      Modify or query the installed toolchains
    target         Modify a toolchain's supported targets
    component      Modify a toolchain's installed components
    override       Modify directory toolchain overrides
    run            Run a command with an environment configured for a given toolchain
    which          Display which binary will be run for a given command
    doc            Open the documentation for the current toolchain
    man            View the man page for a given command
    self           Modify the rustup installation
    set            Alter rustup settings
    completions    Generate tab-completion scripts for your shell
    help           Print this message or the help of the given subcommand(s)

DISCUSSION:
    Rustup installs The Rust Programming Language from the official
    release channels, enabling you to easily switch between stable,
    beta, and nightly compilers and keep them updated. It makes
    cross-compiling simpler with binary builds of the standard library
    for common platforms.

```

</td>
</tr>
</table>

</details>

Reply:
So, I noticed they're using `Command::help_template` there, let's see how it would go after #3205 is merged.